features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java

```diff
@@ -358,4 +369,124 @@ private void checkForDuplicateName (Group group) {
+     * depending on the enabled rules. In order to keep the front-end responsive, reloads are done asynchronously.
+     * Usages of the classification engine are blocked until ongoing reloads did finish. If a reload fails then
+     * future usages of this classification engine also fail until a following reload succeeds.
+     */
+    private static class AsyncReloadingClassificationEngine implements ClassificationEngine {
+
+        private enum State {
+            READY, RELOADING, NEED_ANOTHER_RELOAD, FAILED
+        }
+
+        private final ClassificationEngine delegate;
+
+        // uses at most one additional thread; if the thread is not used for 60 seconds then it is terminated
+        // -> uses no additional resources while being idle
+        private final ExecutorService executorService = new ThreadPoolExecutor(0, 1,
+                60L, TimeUnit.SECONDS,
+                new ArrayBlockingQueue<>(1),
+                runnable -> new Thread(runnable, "AsyncReloadingClassificationEngine")
+        );
+
+        private State state = State.READY;
+        private Exception reloadException;
+
+        public AsyncReloadingClassificationEngine(ClassificationEngine delegate) {
+            this.delegate = delegate;
+        }
+
+        private void setState(State newState) {
+            state = newState;
+            notifyAll();
+        }
+
+        private void waitUntilReadyOrFailed() {
+            while (true) {
+                switch (state) {
+                    case READY: return;
+                    case FAILED: throw new RuntimeException("classification engine can not be used because last reload failed", reloadException);
+                }
+                try {
+                    wait();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        private void doReload() {
+            // this method must not modify the state because it not synchronized
+            try {
+                LOG.debug("reload classification engine");
+                delegate.reload();
+                if (reloadSucceeded()) {
+                    LOG.debug("another classification engine reload is required");
+                    doReload();
+                } else {
+                    LOG.debug("classification engine reloaded");
+                }
+            } catch (Exception e) {
+                LOG.error("reload of classification engine failed", e);
+                if (reloadFailed(e)) {
+                    LOG.debug("another classification engine reload is required");
+                    doReload();
+                } else {
+                    LOG.debug("classification engine reloaded");
+                }
+            }
+        }
+
+        private synchronized boolean reloadSucceeded() {
+            if (state == State.NEED_ANOTHER_RELOAD) {
```

Comment:
The delegate has no state of its own. The delegate is always ready to do classifications. When `reloadSucceded` is called no other reload is in progress. (Ignoring possible background reload triggers; but these have no impact because reloads are idempotent.)